### PR TITLE
My proposal to simplify the `setup`

### PR DESCRIPTION
**Pull request by pberkes**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,54 +1,48 @@
 # -*- coding: utf-8 -*-
-import ast
 import os
+
 from setuptools import setup, find_packages
 
+from grader import (
```
Comment (keszybz) [inline setup.py:6]: Are we sure that we'll always import the right `grader`? E.g. if the user already installed grader in some previous version, will the one from cwd "win"?

File: setup.py [hunk 2]
```diff
@@ -1,54 +1,48 @@
...
-                    pass
-    # return "not available" if we didn't find the variable
-    return 'N/A'
-
-# pick the relevant keywords from the __init__.py file
-metadata_vars = ('name', 'version', 'description', 'author', 'license', 'url')
-metadata = dict((var, parse_keyword('__%s__'%var)) for var in metadata_vars)
-
 # Get the long description from the README file
 with open(os.path.join(CWD, 'README.md'), encoding='utf-8') as f:
```
Comment (keszybz) [inline setup.py:21]: `encoding=utf-8` is the default anyway, can we drop that too?

File: grader/__init__.py [hunk 3]
```diff
@@ -3,31 +3,35 @@
...
+# current dir
+CWD = os.path.abspath(os.path.dirname(__file__))
 
 __name__ =        'grader'
 __description__ = 'a Python module and command-line utility to grade applications'
 __version__ =     '0.1'
 __url__ =         'https://github.com/ASPP/grader'
 __author__ =      'grader contributor'
 __license__ =     'GPLv3+'
-__revision__ =    'N/A'
```
Comment (keszybz) [inline grader/__init__.py:14]: Hmm, but then `__revision__` is not set anymore?

**Review by keszybz — CHANGES_REQUESTED** [review]

File: setup.py (hunk 1 above)
Comment (jni) [reply setup.py:6]: Yes, cwd is the first entry in `sys.path` by default. Test case:

```console
 ~/projects/play/import-order
 $ mkdir numpy
(36)
jni@molino Mon Nov 13 11:41
 ~/projects/play/import-order
 $ echo import os > numpy/__init__.py
(36)
jni@molino Mon Nov 13 11:41
 ~/projects/play/import-order
 $ cat numpy/__init__.py
import os
(36)
jni@molino Mon Nov 13 11:41
 ~/projects/play/import-order
 $ python
Python 3.6.1 |Anaconda custom (x86_64)| (default, May 11 2017, 13:04:09)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> dir(np)
['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'os']
```

Comment (keszybz) [reply setup.py:6]: Indeed. And people almost always run setup.py in the same way: by `cd`-ing into the directory and then calling `python setup.py ...`. So OK.

File: grader/__init__.py (hunk 3 above)
Comment (pberkes) [reply grader/__init__.py:14]: `__revision__` is not used in `setup.py` so it's not a necessary value

if needed a user can call the new `revision()` function and get that number. That avoids launching a subprocess that tries to call `git` every time you import `grader`

Comment (keszybz) [reply grader/__init__.py:14]: Then let's just drop this. It's not like grader has any external users...

Comment (pberkes) [reply grader/__init__.py:14]: just `__revision__` or the `revision()` function as well?

Comment (keszybz) [reply grader/__init__.py:14]: Both, I think. It looks like `__revision__` was added in 335682b46f2ab8bece1ca9c4001a5dfeece8d053 but never used (see `metadata_vars` in that commit).

Comment (otizonaizit) [reply grader/__init__.py:14]: I'd like to have an handle on the git commit the version of grader I am using comes from. As simple as that. That is especially useful if we start tagging grader.conf and store them for later analysis. I don't like the `revision` function, or better, I'd like the functiuon to be run once at `setup.py`, when the grader wheel is created, and then the output of this gets hard-coded in __revision__. This is what `numpy` and `scipy` are doing.

Comment (otizonaizit) [reply grader/__init__.py:14]: To be clear: I don't envision the `__revision__` attribute as something used on a daily basis by a "user" of grader. I expect it to be part of richer metadata for provenance tracking. I agree that this is mostly overkill for grader, but why not try to have the "right" solution? It's not like it requires a lot of over-engineering…

Comment (jni) [reply grader/__init__.py:14]: I agree with @otizonaizit, this is something I always have wanted to know eg. with scikit-image. I'll definitely port this to that project!

**Pull request merged**